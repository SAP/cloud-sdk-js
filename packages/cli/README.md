<!-- sap-cloud-sdk-logo -->
<!-- This block is inserted by scripts/replace-common-readme.ts and not oclif like the commands block. Do not adjust it manually. -->
<a href="https://sap.com/s4sdk"><img src="https://help.sap.com/doc/2324e9c3b28748a4ae2ad08166d77675/1.0/en-US/logo-with-js.svg" alt="SAP Cloud SDK for JavaScript Logo" height="122.92" width="226.773"/></a>
<!-- sap-cloud-sdk-logo-stop -->

# @sap-cloud-sdk/cli

This package contains the command-line interface (CLI) of the SAP Cloud SDK.
The CLI offers commands to scaffold new projects or create typed API clients.

## Installation

```bash
$ npm install @sap-cloud-sdk/cli
```

## Usage

<!-- commands -->
* [`sap-cloud-sdk add-approuter [PROJECTDIR]`](#sap-cloud-sdk-add-approuter-projectdir)
* [`sap-cloud-sdk add-cds [PROJECTDIR]`](#sap-cloud-sdk-add-cds-projectdir)
* [`sap-cloud-sdk add-cx-server [PROJECTDIR]`](#sap-cloud-sdk-add-cx-server-projectdir)
* [`sap-cloud-sdk autocomplete [SHELL]`](#sap-cloud-sdk-autocomplete-shell)
* [`sap-cloud-sdk generate-odata-client`](#sap-cloud-sdk-generate-odata-client)
* [`sap-cloud-sdk help [COMMAND]`](#sap-cloud-sdk-help-command)
* [`sap-cloud-sdk help-page`](#sap-cloud-sdk-help-page)
* [`sap-cloud-sdk init [PROJECTDIR]`](#sap-cloud-sdk-init-projectdir)
* [`sap-cloud-sdk package [PROJECTDIR]`](#sap-cloud-sdk-package-projectdir)

## `sap-cloud-sdk add-approuter [PROJECTDIR]`

Setup your Cloud Foundry app to authenticate through the app router

```
USAGE
  $ sap-cloud-sdk add-approuter [PROJECTDIR]

ARGUMENTS
  PROJECTDIR  Path to the project directory to which the approuter should be added.

OPTIONS
  -h, --help  Show help for the add-approuter command.
  --force     Do not fail if a file already exist and overwrite it.

ALIASES
  $ sap-cloud-sdk add-app-router

EXAMPLE
  $ sap-cloud-sdk add-approuter
```

_See code: [src/commands/add-approuter.ts](https://github.com/SAP/cloud-sdk-js/blob/v1.39.0/src/commands/add-approuter.ts)_

## `sap-cloud-sdk add-cds [PROJECTDIR]`

Setup your Cloud Foundry app to use a CDS service

```
USAGE
  $ sap-cloud-sdk add-cds [PROJECTDIR]

ARGUMENTS
  PROJECTDIR  Path to the project directory in which the cds sources should be added.

OPTIONS
  -h, --help     Show help for the add-cds command.
  -v, --verbose  Show more detailed output.
  --force        Do not fail if a file or npm script already exist and overwrite it.

EXAMPLE
  $ sap-cloud-sdk add-cds
```

_See code: [src/commands/add-cds.ts](https://github.com/SAP/cloud-sdk-js/blob/v1.39.0/src/commands/add-cds.ts)_

## `sap-cloud-sdk add-cx-server [PROJECTDIR]`

Add the scripts to set up a Jenkins server for CI/CD of your project

```
USAGE
  $ sap-cloud-sdk add-cx-server [PROJECTDIR]

ARGUMENTS
  PROJECTDIR  Path to the project directory to which the cx-server should be added.

OPTIONS
  -h, --help  Show help for the add-cx-server command.
  --force     Do not fail if a file already exist and overwrite it.

EXAMPLE
  $ sap-cloud-sdk add-cx-server
```

_See code: [src/commands/add-cx-server.ts](https://github.com/SAP/cloud-sdk-js/blob/v1.39.0/src/commands/add-cx-server.ts)_

## `sap-cloud-sdk autocomplete [SHELL]`

display autocomplete installation instructions

```
USAGE
  $ sap-cloud-sdk autocomplete [SHELL]

ARGUMENTS
  SHELL  shell type

OPTIONS
  -r, --refresh-cache  Refresh cache (ignores displaying instructions)

EXAMPLES
  $ sap-cloud-sdk autocomplete
  $ sap-cloud-sdk autocomplete bash
  $ sap-cloud-sdk autocomplete zsh
  $ sap-cloud-sdk autocomplete --refresh-cache
```

_See code: [@oclif/plugin-autocomplete](https://github.com/oclif/plugin-autocomplete/blob/v0.3.0/src/commands/autocomplete/index.ts)_

## `sap-cloud-sdk generate-odata-client`

Generates a OData client from a edmx service file definition. For SAP solutions, you can find these definitions at https://api.sap.com/.

```
USAGE
  $ sap-cloud-sdk generate-odata-client

OPTIONS
  -i, --inputDir=inputDir                              (required) This directory will be recursively searched for
                                                       .edmx/.xml files.

  -n, --processesJsGeneration=processesJsGeneration    [default: 16] Number of processes used for generation of
                                                       javascript files.

  -o, --outputDir=outputDir                            (required) Directory to save the generated code in.

  -s, --serviceMapping=serviceMapping                  Configuration file to ensure consistent names between multiple
                                                       generation runs with updated / changed metadata files. Will be
                                                       generated if not existent. By default it will be saved to/read
                                                       from the input directory as "service-mapping.json".

  --additionalFiles=additionalFiles                    Glob describing additional files to be added to the each
                                                       generated service directory.

  --aggregatorDirectoryName=aggregatorDirectoryName    Hack for cloud-sdk-vdm package

  --aggregatorNpmPackageName=aggregatorNpmPackageName  When provided, the generator will generate an additional package
                                                       with the provided name that has dependencies to all other
                                                       generated packages.

  --clearOutputDir                                     When set to true, the generator will delete EVERYTHING in the
                                                       specified output directory before generating code. [default:
                                                       false].

  --forceOverwrite                                     By default, the generator will exit when encountering a file that
                                                       already exists. When set to true, it will be overwritten instead.
                                                       Please note that compared to the --clearOutputDir option, this
                                                       will not delete outdated files. [default: false].

  --generateCSN                                        When set to true a CSN file will be generated for each service
                                                       definition in the output directory. [default: false].

  --[no-]generateJs                                    By default, the generator will also generate transpiled .js,
                                                       .js.map, .d.ts and .d.ts.map files. When set to false, the
                                                       generator will only generate .ts files. [default: true].

  --[no-]generatePackageJson                           By default, the generator will generate a package.json file,
                                                       specifying dependencies and scripts for compiling and generating
                                                       documentation. When set to false, the generator will skip the
                                                       generation of the package.json. [default: true].

  --[no-]generateTypedocJson                           By default, the generator will generate a typedoc.json file for
                                                       each package, used for the corresponding "doc" npm script. When
                                                       set to false, the generator will skip the generation of the
                                                       typedoc.json. [default: true].

  --projectDir=projectDir                              [default: .] Path to the folder in which the VDM should be
                                                       created. The input and output dir are relative to this directory.

  --s4hanaCloud                                        When set to true, the description of the generated packages will
                                                       be specific to S/4HANA Cloud. [default: false].

  --sdkAfterVersionScript                              When set to true, the package.json of generated services will
                                                       have the after-version script to internally keep the versions in
                                                       sync. [default: false].

  --useSwagger                                         Augment parsed information with information from swagger
                                                       definition files. Files are expected to have the same name as the
                                                       edmx file, but with .json as suffix. [default: false].

  --versionInPackageJson=versionInPackageJson          By default, when generating package.json file, the generator will
                                                       set a version by using the generator version. It can also be set
                                                       to a specific version.

  --writeReadme                                        When set to true, the generator will write a README.md file into
                                                       the root folder of every package. This option does not make that
                                                       much sense without also set useSwagger to "true". [default:
                                                       false].

EXAMPLES
  $ sap-cloud-sdk generate-odata-client -i directoryWithEdmxFiles -o outputDirectory --forceOverwrite
  $ sap-cloud-sdk generate-odata-client --help
```

_See code: [src/commands/generate-odata-client.ts](https://github.com/SAP/cloud-sdk-js/blob/v1.39.0/src/commands/generate-odata-client.ts)_

## `sap-cloud-sdk help [COMMAND]`

display help for sap-cloud-sdk

```
USAGE
  $ sap-cloud-sdk help [COMMAND]

ARGUMENTS
  COMMAND  command to show help for

OPTIONS
  --all  see all commands in CLI
```

_See code: [@oclif/plugin-help](https://github.com/oclif/plugin-help/blob/v3.2.2/src/commands/help.ts)_

## `sap-cloud-sdk help-page`

Display the product page, which contains tutorials and links to all relevant resources

```
USAGE
  $ sap-cloud-sdk help-page
```

_See code: [src/commands/help-page.ts](https://github.com/SAP/cloud-sdk-js/blob/v1.39.0/src/commands/help-page.ts)_

## `sap-cloud-sdk init [PROJECTDIR]`

Initializes your project for the SAP Cloud SDK, SAP Cloud Platform Cloud Foundry and CI/CD using the SAP Cloud SDK toolkit

```
USAGE
  $ sap-cloud-sdk init [PROJECTDIR]

ARGUMENTS
  PROJECTDIR  Path to the directory in which the project should be created.

OPTIONS
  -h, --help               Show help for the init command.
  -v, --verbose            Show more detailed output.

  --addCds                 Add a cds configuration and example data to follow the SAP Cloud Application Programming
                           model.

  --force                  Do not fail if a file or npm script already exist and overwrite it.

  --frontendScripts        Add frontend-related npm scripts which are executed by our CI/CD toolkit.

  --projectDir=projectDir  Path to the directory in which the project should be created.

EXAMPLES
  $ sap-cloud-sdk init
  $ sap-cloud-sdk init --help
```

_See code: [src/commands/init.ts](https://github.com/SAP/cloud-sdk-js/blob/v1.39.0/src/commands/init.ts)_

## `sap-cloud-sdk package [PROJECTDIR]`

Copies the specified files to the deployment folder

```
USAGE
  $ sap-cloud-sdk package [PROJECTDIR]

ARGUMENTS
  PROJECTDIR  Path to the project directory that shall be packaged.

OPTIONS
  -e, --exclude=exclude  Comma separated list of files or globs to exclude
  -h, --help             Show help for the package command.

  -i, --include=include  [default: package.json,package-lock.json,index.js,dist/**/*] Comma seperated list of files or
                         globs to include

  -o, --output=output    [default: deployment] Output and deployment folder

  -v, --verbose          Show more detailed output.

  --ci                   Add node_modules in production environments to respect the `build once` principle.

EXAMPLES
  $ sap-cloud-sdk package
  $ sap-cloud-sdk package -i="index.html"
  $ sap-cloud-sdk package --include="package.json,package-lock.json,index.js,dist/**/*" --exclude="**/*.java"
```

_See code: [src/commands/package.ts](https://github.com/SAP/cloud-sdk-js/blob/v1.39.0/src/commands/package.ts)_
<!-- commandsstop -->

<!-- sap-cloud-sdk-common-readme -->
<!-- This block is inserted by scripts/replace-common-readme.ts and not oclif like the commands block. Do not adjust it manually. -->
## Support

The recommended way to get in touch with us is to create an issue in our [github repository](https://github.com/SAP/cloud-sdk-js/issues).
Select the issue category `Bug`, `Feature` or `Question` depending on the nature of your request.
We try to provide fixes, features and answers as soon as possible.

We also monitor questions on [StackOverflow](https://stackoverflow.com/questions/tagged/sap-cloud-sdk?tab=Newest) and [ansers.sap.com](https://answers.sap.com/tags/73555000100800000895) but prefer issues on github.

## Contribute

If you would like to contribute to the SAP Cloud SDK, please make yourself familiar with our [contributing guidelines](https://github.com/SAP/cloud-sdk-js/blob/main/CONTRIBUTING.md) and follow the given instructions.

## Links

- [Github](https://github.com/SAP/cloud-sdk-js)
- [Github - Releases](https://github.com/SAP/cloud-sdk-js/releases)

<br>

- [SAP Cloud SDK Documentation portal](https://sap.github.io/cloud-sdk/)
- [SAP Cloud SDK Documentation portal - Getting started guide](https://sap.github.io/cloud-sdk/docs/js/getting-started)
- [SAP Cloud SDK Documentation portal - API documentation](https://sap.github.io/cloud-sdk/docs/js/api-reference-js-ts)

<br>

- [developers.sap.com - Product Overview](https://developers.sap.com/topics/cloud-sdk.html)
- [developers.sap.com - Tutorials](https://developers.sap.com/tutorial-navigator.html?tag=products:technology-platform/sap-cloud-sdk/sap-cloud-sdk&tag=topic:javascript)

## License

The SAP Cloud SDK is released under the [Apache License Version 2.0.](http://www.apache.org/licenses/)
<!-- sap-cloud-sdk-common-readme-stop -->
