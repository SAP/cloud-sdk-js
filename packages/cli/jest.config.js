// eslint-disable-next-line
const commonConfig = require('../../test-resources/jest.common.config');
module.exports = {
  ...commonConfig,
  displayName: 'cli',
  testPathIgnorePatterns: ['/test/test-output/','/test/nest','/test/express'],
  setupFilesAfterEnv: ['jest-extended']
};
