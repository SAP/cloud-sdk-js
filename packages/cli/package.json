{
  "name": "@sap-cloud-sdk/cli",
  "version": "1.37.0",
  "description": "SAP Cloud SDK for JavaScript cli",
  "homepage": "https://sap.github.io/cloud-sdk/docs/js/overview-cloud-sdk-for-javascript",
  "license": "Apache-2.0",
  "private": true,
  "keywords": [
    "sap-cloud-sdk",
    "cloud-sdk",
    "sap-cloud-platform",
    "cli"
  ],
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "publishConfig": {
    "access": "private"
  },
  "files": [
    "dist/**/*.js",
    "dist/**/*.js.map",
    "dist/**/*.d.ts",
    "dist/**/*.d.ts.map"
  ],
  "repository": "github:SAP/cloud-sdk-js",
  "scripts": {
    "compile": "yarn tsc -b",
    "prepare": "yarn compile",
    "test": "yarn jest --coverage",
    "check:dependencies": "depcheck . --ignores=@oclif/plugin-*"
  },
  "dependencies": {
    "@oclif/command": "^1.5.19",
    "@oclif/config": "^1.14.0",
    "@oclif/parser": "^3.8.4",
    "@oclif/plugin-autocomplete": "^0.3.0",
    "@oclif/plugin-help": "^3.1.0",
    "@oclif/plugin-not-found": "^1.2.3",
    "@oclif/plugin-warn-if-update-available": "^1.7.0",
    "cli-ux": "^5.4.5",
    "execa": "^5.0.0",
    "fast-glob": "^3.2.2",
    "handlebars": "^4.7.3",
    "js-yaml": "^4.0.0",
    "listr": "^0.14.3",
    "rimraf": "^3.0.2",
    "yargs": "16.2.0"
  },
  "devDependencies": {
    "@types/fs-extra": "^9.0.1",
    "@types/js-yaml": "^3.12.3",
    "@types/listr": "^0.14.2",
    "@types/node": ">=13.9.8",
    "@types/rimraf": "^3.0.0",
    "fs-extra": "^9.0.0",
    "jest": "^26.6.1",
    "jest-extended": "^0.11.5",
    "ts-jest": "^26.4.3",
    "typescript": "^4.0.3"
  },
  "oclif": {
    "commands": "./lib/commands",
    "warn-if-update-available": {
      "timeoutInDays": 7
    },
    "macos": {
      "identifier": "com.sap.cloud-sdk-cli"
    },
    "bin": "sap-cloud-sdk",
    "plugins": [
      "@oclif/plugin-help",
      "@oclif/plugin-autocomplete",
      "@oclif/plugin-not-found",
      "@oclif/plugin-warn-if-update-available"
    ]
  }
}
