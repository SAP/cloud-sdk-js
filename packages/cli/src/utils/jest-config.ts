/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { readFileSync, writeFileSync } from 'fs';
import { recordWarning } from './warnings';

export function getJestConfig(isUnitTests: boolean) {
  return {
    reporters: [
      'default',
      [
        'jest-junit',
        {
          suiteName: 'backend unit tests',
          outputDirectory: `./s4hana_pipeline/reports/backend-${
            isUnitTests ? 'unit' : 'integration'
          }`
        }
      ]
    ],
    collectCoverage: true,
    coverageReporters: ['text', 'cobertura'],
    coverageDirectory: `../s4hana_pipeline/reports/coverage-reports/backend-${
      isUnitTests ? 'unit' : 'integration'
    }`
  };
}

export function modifyJestConfig(jestConfigPath: string, data: any) {
  try {
    const jestConfig = JSON.parse(
      readFileSync(jestConfigPath, {
        encoding: 'utf8'
      })
    );
    const adjustedJestConfig = {
      ...jestConfig,
      ...data
    };

    writeFileSync(jestConfigPath, JSON.stringify(adjustedJestConfig, null, 2));
  } catch (error) {
    recordWarning(
      `Could not edit your Jest config at "${jestConfigPath}".`,
      'Please verify if the location is correct and consider opening a bug ticket',
      'at https://github.com/SAP/cloud-sdk-cli'
    );
  }
}
