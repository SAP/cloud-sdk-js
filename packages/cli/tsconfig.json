{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "rootDir": "./src",
    "outDir": "./dist",
    "types": ["node", "jest", "jest-extended"],
    "tsBuildInfoFile": "./dist/.tsbuildinfo",
    "composite": true,
    "lib": ["dom"],
    "target": "es2017"
  },
  "include": ["src/**/*.ts"],
  "exclude": ["dist/**/*", "test/**/*", "**/*.spec.ts", "node_modules/**/*"],
  "files": [],
  "references": []
}
