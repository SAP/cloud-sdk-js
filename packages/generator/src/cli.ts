#!/usr/bin/env node

import { createLogger, ErrorWithCause } from '@sap-cloud-sdk/util';
import { parseCmdArgs } from './cli-parser';
import { generate } from './generator';

const logger = createLogger({
  package: 'generator',
  messageContext: 'cli'
});

generate(parseCmdArgs(process.argv.slice(2)))
  .then(() => logger.info('Generation of services finished successfully.'))
  .catch(err => {
    logger.error(new ErrorWithCause('Generation of services failed.', err));
    process.exit(1);
  });
