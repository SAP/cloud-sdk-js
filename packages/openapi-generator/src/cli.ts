#!/usr/bin/env node

import { createLogger } from '@sap-cloud-sdk/util';
import yargs from 'yargs';
// eslint-disable-next-line import/no-internal-modules
import { hideBin } from 'yargs/helpers';
import { parseOptionsFromConfig, getSpecifiedFlags, cli } from './options';
import { generateWithParsedOptions } from './generator';

const logger = createLogger('openapi-generator');

parseCmdArgs();
/**
 * @internal
 */
export default async function parseCmdArgs(): Promise<void> {
  try {
    const argv = await cli(process.argv).argv;

    if (argv.config) {
      await generateWithParsedOptions({
        ...(await parseOptionsFromConfig(argv.config)),
        ...getSpecifiedFlags(
          argv,
          Object.keys(await yargs(hideBin(process.argv)).argv)
        )
      });
    } else {
      await generateWithParsedOptions(argv);
    }
  } catch (err) {
    logger.error(err);
    yargs.exit(1, new Error());
  }
}
