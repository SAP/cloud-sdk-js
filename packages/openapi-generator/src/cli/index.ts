/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { resolve } from 'path';
import { createLogger } from '@sap-cloud-sdk/util';
import { Command, flags } from '@oclif/command';
import cli from 'cli-ux';
import { generate } from '../generator';

const logger = createLogger('openapi-generator');

class GenerateOpenApiClient extends Command {
  static description =
    'Generate OpenAPI clients, that use the connectivity features of the SAP Cloud SDK for JavaScript.';

  static usage = '--inputDir <inputDirectory> --outputDir <outputDirectory>';

  // eslint-disable-next-line @typescript-eslint/no-var-requires
  static version = require('../../package.json').version;

  static flags = {
    input: flags.string({
      name: 'input',
      char: 'i',
      description:
        'Input directory or file for the OpenAPI service definitions.',
      parse: input => resolve(input),
      required: true
    }),
    outputDir: flags.string({
      name: 'outputDir',
      char: 'o',
      description: 'Output directory for the generated OpenAPI client.',
      parse: input => resolve(input),
      required: true
    }),
    clearOutputDir: flags.boolean({
      name: 'clearOutputDir',
      description:
        'Remove all files in the output directory before generation.',
      default: false,
      required: false
    }),
    strictNaming: flags.boolean({
      name: 'strictNaming',
      description:
        'By default, the generation fails, when there are duplicate names for operations and/or path parameters after transforming them to camel case. Set this to true to enable unique name generation. The names will then be generated by appending numbers.',
      default: true,
      required: false,
      allowNo: true
    }),
    versionInPackageJson: flags.string({
      name: 'versionInPackageJson',
      description:
        'By default, when generating package.json file, the generator will set a version by using the generator version. It can also be set to a specific version.',
      required: false
    }),
    generatePackageJson: flags.boolean({
      name: 'generatePackageJson',
      description:
        'By default, the generator will generate a package.json file, specifying dependencies and scripts for compiling and generating documentation. When set to false, the generator will skip the generation of the package.json.',
      default: true,
      required: false
    }),
    generateJs: flags.boolean({
      name: 'generateJs',
      description:
        'By default, the generator will also generate transpiled .js, .js.map, .d.ts and .d.ts.map files. When set to false, the generator will only generate .ts files.',
      default: true,
      required: false
    }),
    serviceMapping: flags.string({
      name: 'serviceMapping',
      description:
        'Configuration file to ensure consistent names between multiple generation runs with updated / changed metadata files. By default it will be read from the input directory as "service-mapping.json".',
      parse: input => resolve(input),
      required: false
    }),
    tsConfig: flags.string({
      name: 'tsConfig',
      description:
        'tsconfig.json file to overwrite the default "tsconfig.json".',
      parse: input => resolve(input),
      required: false
    }),
    additionalFiles: flags.string({
      name: 'additionalFiles',
      description:
        'Glob describing additional files to be added to the each generated service directory.',
      parse: input => resolve(input),
      required: false,
      hidden: true
    }),
    writeReadme: flags.boolean({
      name: 'writeReadme',
      description:
        'When set to true, the generator will write a README.md file into the root folder of every package.',
      default: false,
      required: false,
      hidden: true
    })
  };

  async run(): Promise<void> {
    try {
      const parsed = this.parse(GenerateOpenApiClient);
      await generate(parsed.flags);
    } catch (e) {
      logger.error(e.message);
      return cli.exit(1);
    }
  }
}

export = GenerateOpenApiClient;
