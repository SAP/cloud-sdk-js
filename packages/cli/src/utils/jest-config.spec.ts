/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
import { resolve } from 'path';
import fs from 'fs-extra';
import {
  deleteAsync,
  getCleanProjectDir,
  getTestOutputDir,
  TimeThresholds
} from '../../test/test-utils';
import { testDir } from '../../../../test-resources/cli';
import { getJestConfig, modifyJestConfig } from './jest-config';

const testOutputDir = getTestOutputDir(__filename);

describe('Jest Config Utils', () => {
  beforeAll(async () => {
    await deleteAsync(testOutputDir, 3);
  }, TimeThresholds.EXTRA_SHORT);

  it(
    'returns jest config object for tests',
    () => {
      expect(getJestConfig(true)).toMatchSnapshot();
      expect(getJestConfig(false)).toMatchSnapshot();
    },
    TimeThresholds.EXTRA_SHORT
  );

  it(
    'returns adds jest config to file',
    async () => {
      const projectDir = await getCleanProjectDir(testOutputDir, 'jest-config');
      const jestConfigPath = resolve(projectDir, 'jest-e2e.json');
      await fs.copyFile(
        resolve(testDir, 'nest', 'test', 'jest-e2e.json'),
        jestConfigPath
      );

      modifyJestConfig(jestConfigPath, getJestConfig(false));

      expect(
        JSON.parse(await fs.readFile(jestConfigPath, { encoding: 'utf8' }))
      ).toMatchObject(getJestConfig(false));
    },
    TimeThresholds.EXTRA_SHORT
  );
});
