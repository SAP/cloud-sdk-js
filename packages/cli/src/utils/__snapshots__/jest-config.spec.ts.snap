// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jest Config Utils returns jest config object for tests 1`] = `
Object {
  "collectCoverage": true,
  "coverageDirectory": "../s4hana_pipeline/reports/coverage-reports/backend-unit",
  "coverageReporters": Array [
    "text",
    "cobertura",
  ],
  "reporters": Array [
    "default",
    Array [
      "jest-junit",
      Object {
        "outputDirectory": "./s4hana_pipeline/reports/backend-unit",
        "suiteName": "backend unit tests",
      },
    ],
  ],
}
`;

exports[`Jest Config Utils returns jest config object for tests 2`] = `
Object {
  "collectCoverage": true,
  "coverageDirectory": "../s4hana_pipeline/reports/coverage-reports/backend-integration",
  "coverageReporters": Array [
    "text",
    "cobertura",
  ],
  "reporters": Array [
    "default",
    Array [
      "jest-junit",
      Object {
        "outputDirectory": "./s4hana_pipeline/reports/backend-integration",
        "suiteName": "backend unit tests",
      },
    ],
  ],
}
`;
